contracts/TwabRewards.sol

```diff
@@ -0,0 +1,422 @@
+
+    /* ============ Modifiers ============ */
+
+    /// @dev Ensure that the caller is the creator of the currently active promotion.
+    /// @param _promotionId Id of the promotion to check
+    modifier onlyPromotionCreator(uint256 _promotionId) {
+        require(
+            msg.sender == _getPromotion(_promotionId).creator,
+            "TwabRewards/only-promotion-creator"
+        );
+        _;
+    }
+
+    /* ============ External Functions ============ */
+
+    /// @inheritdoc ITwabRewards
+    function createPromotion(
+        address _ticket,
+        address _token,
+        uint216 _tokensPerEpoch,
+        uint32 _startTimestamp,
+        uint32 _epochDuration,
+        uint8 _numberOfEpochs
+    ) external override returns (uint256) {
+        _requireTicket(_ticket);
+
+        uint256 _nextPromotionId = _latestPromotionId + 1;
+        _latestPromotionId = _nextPromotionId;
+
+        IERC20(_token).safeTransferFrom(
+            msg.sender,
+            address(this),
+            _tokensPerEpoch * _numberOfEpochs
+        );
+
+        Promotion memory _nextPromotion = Promotion(
+            msg.sender,
+            _ticket,
+            _token,
+            _tokensPerEpoch,
+            _startTimestamp,
+            _epochDuration,
+            _numberOfEpochs
+        );
+
+        _promotions[_nextPromotionId] = _nextPromotion;
+
+        emit PromotionCreated(_nextPromotionId);
+
+        return _nextPromotionId;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function cancelPromotion(uint256 _promotionId, address _to)
+        external
+        override
+        onlyPromotionCreator(_promotionId)
+        returns (bool)
+    {
+        require(_isPromotionActive(_promotionId) == true, "TwabRewards/promotion-not-active");
+        require(_to != address(0), "TwabRewards/recipient-not-zero-address");
+
+        Promotion memory _promotion = _getPromotion(_promotionId);
+
+        IERC20 _token = IERC20(_promotion.token);
+        uint256 _remainingRewards = _getRemainingRewards(_promotionId);
+
+        if (_remainingRewards > 0) {
+            _token.safeTransfer(_to, _remainingRewards);
+        }
+
+        delete _promotions[_promotionId];
+
+        emit PromotionCancelled(_promotionId, _remainingRewards);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function extendPromotion(uint256 _promotionId, uint8 _numberOfEpochs)
+        external
+        override
+        returns (bool)
+    {
+        require(_isPromotionActive(_promotionId) == true, "TwabRewards/promotion-not-active");
+
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint8 _extendedNumberOfEpochs = _promotion.numberOfEpochs + _numberOfEpochs;
+
+        _requireEpochLimit(_extendedNumberOfEpochs);
+
+        IERC20 _token = IERC20(_promotion.token);
+        uint256 _amount = _numberOfEpochs * _promotion.tokensPerEpoch;
+
+        _token.safeTransfer(address(this), _amount);
```

Comment:
CEI move this one line lower

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/0bde18efe9ed24761b7289f6ffe9dd64d245e38c